src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -602,20 +602,35 @@ public bool WriteIndented
                     new ReflectionEmitCachingMemberAccessor() :
                     new ReflectionMemberAccessor();
 #elif NETFRAMEWORK
-                new ReflectionEmitCachingMemberAccessor();
+                    new ReflectionEmitCachingMemberAccessor();
 #else
-                new ReflectionMemberAccessor();
+                    new ReflectionMemberAccessor();
 #endif
 
-        internal bool IsImmutable
+        /// <summary>
+        /// Specifies whether the current instance has been locked for modification.
+        /// </summary>
+        /// <remarks>
+        /// A <see cref="JsonSerializerOptions"/> instance can be locked either if
+        /// it has been passed to one of the <see cref="JsonSerializer"/> methods,
+        /// has been associated with a <see cref="JsonSerializerContext"/> instance,
+        /// or a user explicitly called the <see cref="MakeReadOnly"/> method on the instance.
+        /// </remarks>
+        public bool IsReadOnly => _isReadOnly;
+
+        /// <summary>
+        /// Locks the current instance for further modification.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The instance does not specify a <see cref="TypeInfoResolver"/> setting.</exception>
+        /// <remarks>This method is idempotent.</remarks>
+        public void MakeReadOnly()
         {
-            get => _isImmutable;
-            set
+            if (_typeInfoResolver is null)
             {
-                Debug.Assert(value, "cannot unlock options instances");
-                Debug.Assert(_typeInfoResolver != null, "cannot lock without a resolver.");
-                _isImmutable = true;
+                ThrowHelper.ThrowInvalidOperationException_JsonSerializerOptionsNoTypeInfoResolverSpecified();
```

Comment:
should this be assigning default resolver instead? I guess we can re-iterate on that in the future if we get feedback but I'd prefer more intuitive behavior than linker safety in this case

Reply:
No -- we would need to mark the method as linker-unsafe, and I'd rather avoid doing that in cases other than ones mandated by backward compatibility.